plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/frontend.tsx

```diff
@@ -38,205 +39,136 @@ interface Store {
-			}
-			return state.numberOfItemsInTheCart > 0;
+const { state: wooState } = store< WooStore >( 'woocommerce' );
+
+// Todo: Remove the type cast once we import from `@wordpress/interactivity`.
+const { state } = ( store as typeof StoreType )< Store >(
+	'woocommerce/product-button',
+	{
+		state: {
+			get quantity() {
+				const { productId } = getContext();
+				const product = wooState.cart.items.find(
+					( item ) => item.id === productId
+				);
+				return product?.quantity || 0;
+			},
+			get slideInAnimation() {
+				const { animationStatus } = getContext();
+				return animationStatus === AnimationStatus.SLIDE_IN;
+			},
+			get slideOutAnimation() {
+				const { animationStatus } = getContext();
+				return animationStatus === AnimationStatus.SLIDE_OUT;
+			},
+			get hasCartLoaded(): boolean {
+				return !! wooState.cart;
+			},
+			get addToCartText(): string {
+				const { animationStatus, tempQuantity, addToCartText } =
+					getContext();
+
+				// We use the temporary quantity when there's no animation, or when
+				// the second part of the animation hasn't started yet.
+				const showTemporaryNumber =
+					animationStatus === AnimationStatus.IDLE ||
+					animationStatus === AnimationStatus.SLIDE_OUT;
+				const quantity = showTemporaryNumber
+					? tempQuantity
+					: state.quantity;
+
+				if ( quantity === 0 ) return addToCartText;
+				return state.inTheCartText.replace(
+					'###',
+					quantity.toString()
+				);
+			},
+			get displayViewCart(): boolean {
+				const { displayViewCart, tempQuantity } = getContext();
+				if ( ! displayViewCart ) return false;
+				if ( ! state.hasCartLoaded ) {
+					return tempQuantity > 0;
+				}
+				return state.quantity > 0;
+			},
 		},
-	},
-	actions: {
-		*addToCart() {
-			const context = getContext();
-			const { productId, quantityToAdd } = context;
+		actions: {
+			*addToCart() {
+				const context = getContext();
+				const { productId, quantityToAdd } = context;
 
-			context.isLoading = true;
+				// Todo: move the addToCart store to its own module.
+				const { actions } = ( yield import(
+					'../../../../base/stores/add-to-cart'
+				) ) as WooStore;
 
-			try {
-				yield dispatch( storeKey ).addItemToCart(
+				actions.addToCart(
 					productId,
-					quantityToAdd
+					// Question: is quantityToAdd available in the cart or we need to pass it down?
+					state.quantity + quantityToAdd
 				);
-
-				// After the cart is updated, sync the temporary number of items again.
-				context.temporaryNumberOfItems = state.numberOfItemsInTheCart;
-			} catch ( error ) {
-				const storeNoticeBlock =
-					document.querySelector( storeNoticeClass );
-
-				if ( ! storeNoticeBlock ) {
-					document
-						.querySelector( '.entry-content' )
-						?.prepend( createNoticeContainer() );
+			},
+			*refreshCart() {
+				// Question: use `splitTask` here? Does it make sense?
+				// Question: does it make sense to use a dynamic import instead of a static one?
```

Comment:
Question: does it make sense to use a dynamic import instead of a static one?

Reply:
For now, I'm going to leave it as a dynamic import, but I still want to think about whether it makes sense to avoid preloading this module.